ldms/src/sampler/llnl/lustre_client_general.c

```diff
@@ -0,0 +1,276 @@
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdio.h>
+#include <stdint.h>
+#include <dirent.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "ldms.h"
+#include "ldmsd.h"
+#include "ldms_jobid.h"
+#include "lustre_client.h"
+#include "lustre_client_general.h"
+
+static ldms_schema_t llite_general_schema;
+
+/* all of these will be prefixed with "stats." in the schema */
+static char *llite_stats_uint64_t_entries[] = {
+        "dirty_pages_hits",
+        "dirty_pages_misses",
+        "read_bytes.sum",
+        "write_bytes.sum",
+        "brw_read.sum",
+        "brw_write.sum",
+        "ioctl",
+        "open",
+        "close",
+        "mmap",
+        "page_fault",
+        "page_mkwrite",
+        "seek",
+        "fsync",
+        "readdir",
+        "setattr",
+        "truncate",
+        "flock",
+        "getattr",
+        "create",
+        "link",
+        "unlink",
+        "symlink",
+        "mkdir",
+        "rmdir",
+        "mknod",
+        "rename",
+        "statfs",
+        "alloc_inode",
+        "setxattr",
+        "getxattr",
+        "getxattr_hits",
+        "listxattr",
+        "removexattr",
+        "inode_permission",
+        NULL
+};
+
+int llite_general_schema_is_initialized()
+{
+        if (llite_general_schema != NULL)
+                return 0;
+        else
+                return -1;
+}
+
+int llite_general_schema_init()
+{
+        ldms_schema_t sch;
+        int rc;
+        int i;
+
+        log_fn(LDMSD_LDEBUG, SAMP" llite_general_schema_init()\n");
+        sch = ldms_schema_new("lustre_client");
+        if (sch == NULL)
+                goto err1;
+        rc = ldms_schema_meta_array_add(sch, "hostname", LDMS_V_CHAR_ARRAY, 64);
+        if (rc < 0)
+                goto err2;
+        rc = ldms_schema_meta_array_add(sch, "fs_name", LDMS_V_CHAR_ARRAY, 64);
+        if (rc < 0)
+                goto err2;
+        rc = ldms_schema_meta_array_add(sch, "llite", LDMS_V_CHAR_ARRAY, 64);
+        if (rc < 0)
+                goto err2;
+	rc = ldms_schema_meta_add(sch, "component_id", LDMS_V_U64);
+        if (rc < 0)
+                goto err2;
+	rc = LJI_ADD_JOBID(sch);
```

Comment:
save the rc value from LJI_ADD_JOBID somewhere and use it rather than ldms_metric_by_name.